iceoryx_posh/source/popo/ports/subscriber_port_multi_producer.cpp

```diff
@@ -26,15 +26,64 @@ SubscriberPortMultiProducer::SubscriberPortMultiProducer(
 
 cxx::optional<capro::CaproMessage> SubscriberPortMultiProducer::getCaProMessage() noexcept
 {
-    /// @todo
-    return cxx::nullopt_t();
+    // get subscribe request from user side
+    const auto currentSubscribeRequest = getMembers()->m_subscribeRequested.load(std::memory_order_relaxed);
+
+    const auto currentSubscriptionState = getMembers()->m_subscriptionState.load(std::memory_order_relaxed);
+
+    if (currentSubscribeRequest && (SubscribeState::NOT_SUBSCRIBED == currentSubscriptionState))
+    {
+        getMembers()->m_subscriptionState.store(SubscribeState::WAIT_FOR_OFFER, std::memory_order_relaxed);
+
+        capro::CaproMessage caproMessage(capro::CaproMessageType::SUB, BasePort::getMembers()->m_serviceDescription);
+        caproMessage.m_chunkQueueData = static_cast<void*>(&getMembers()->m_chunkReceiverData);
+        caproMessage.m_historyCapacity = getMembers()->m_historyRequest;
+
+        return cxx::make_optional<capro::CaproMessage>(caproMessage);
+    }
+    else if (!currentSubscribeRequest && (SubscribeState::WAIT_FOR_OFFER == currentSubscriptionState))
+    {
+        getMembers()->m_subscriptionState.store(SubscribeState::NOT_SUBSCRIBED, std::memory_order_relaxed);
+
+        capro::CaproMessage caproMessage(capro::CaproMessageType::UNSUB, BasePort::getMembers()->m_serviceDescription);
+        caproMessage.m_chunkQueueData = static_cast<void*>(&getMembers()->m_chunkReceiverData);
+
+        return cxx::make_optional<capro::CaproMessage>(caproMessage);
+    }
+    else
+    {
+        // nothing to change
+        return cxx::nullopt_t();
+    }
 }
 
 cxx::optional<capro::CaproMessage>
-SubscriberPortMultiProducer::dispatchCaProMessage(const capro::CaproMessage& caProMessage [[gnu::unused]]) noexcept
+SubscriberPortMultiProducer::dispatchCaProMessage(const capro::CaproMessage& caProMessage[[gnu::unused]]) noexcept
 {
-    /// @todo
-    return cxx::nullopt_t();
+    const auto currentSubscriptionState = getMembers()->m_subscriptionState.load(std::memory_order_relaxed);
+
+    if ((capro::CaproMessageType::OFFER == caProMessage.m_type)
+        && (SubscribeState::WAIT_FOR_OFFER == currentSubscriptionState))
+    {
+        capro::CaproMessage caproMessage(capro::CaproMessageType::SUB, BasePort::getMembers()->m_serviceDescription);
+        caproMessage.m_chunkQueueData = static_cast<void*>(&getMembers()->m_chunkReceiverData);
+        caproMessage.m_historyCapacity = getMembers()->m_historyRequest;
+
+        return cxx::make_optional<capro::CaproMessage>(caproMessage);
+    }
+    else if ((capro::CaproMessageType::ACK == caProMessage.m_type)
+             || (capro::CaproMessageType::NACK == caProMessage.m_type)
+             || (capro::CaproMessageType::STOP_OFFER == caProMessage.m_type))
+    {
+        // we ignore all these messages for multi-producer
```

Comment:
If we ignore all those messages then where is the subscription state updated? I think at the moment this class has only the state WAIT_FOR_OFFER or NOT_SUBSCRIBED. Is this intended? 

I think for the multi producer subscriber we need a list of all the producer to which the subscriber is connected. When we receive a STOP_OFFER from them we have to cross them from our list and if no more producers are available we switch to the NOT_SUBSCRIBED state.

We could also go in the direction of ignoring the subscription state but then the multi producer subscriber port would behave differently and then the question arises: is this something a user would expect from such a subscriber?

Reply:
We keep it simple first. In an n:m pub/sub setup the subscription state is no more easy. When are you subscribed, when there is at least one publisher. Or maybe there are two and you would need both to be fine, would you then be subscribed if you are subscribed to all the ones you have in mind? A bookkeeping as you propose is a lot effort and I would first get some experience and see if we want to go for this. I also have the feeling that in other n:m technologies like DDS there is no subscription state like this for a subscriber. They more look on available publishers and subscribers on a topic.  WAIT_FOR_OFFER was maybe not the best choice. I would propose that in a multi-pusher setup we have NOT_SUBSCRIBED and SUBSCRIBED. And this state reflects the subscriber wish. I.e. no matter how many publishers there are, if you want to be subscribed we subscribe to all available ones and are in the SUBSCRIBED state. We can make it more complex later if we see the need. What do you think @elfenpiff

Reply:
I agree!